SHS-6201: Implement: Editable Metatags

nJim opened the pull request:
# Summary
- Creates a new field to hold a metatag image
- Moves this field to the metatags field group in the sidebar of the node form
- Set image related metatags to use the value of this image field
- Updates previous logic to use field_metatags_image before falling back on field_hs_page_hero media

This PR also includes previous work that was organized in two other PRs but is now a part of this changeset. Other changes include:
- Improves default values for metatags by including a new optimized image style and alt text for media
- Adding new fields for field_metatags and field_metatags_image for the flexible page
- Adding and enabling the metatag_simple_widget module
- Enabling the metatag module for twitter fields (metatag_twitter_cards)

## Backstory
This is part of an initiative to allow authors better control of their content metadata. This work introduces a new metatag widget to simply editorial controls. It includes better helptext and a focused set of input that control the basics needed for pages.

## Need Review By (Date)
Not set.

## Urgency
Not set.

## Steps to Test
** Note: This test plan is for all three tickets in this series. There is no need to test them separately. **

1. [Login to testing environment](https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com/user) as a user that can edit content.
2. [Edit the editing 'Short Spotlight(s) - Classic' node](https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com/node/18626/edit).
3. Test meta field form
  a. Locate the 'Metatags' section in the advanced section on the right side of the form. Verify the field instructions, controls, and help text match the attached screenshot. 
  b. Set the page title to 'My page', the meta title to 'My custom title' and meta description to 'Here is my custom summary'. Save the node.
  c. Locate the 'Meta Image' field. Verify this is a media reference field, limited to 1 item, no required, restricted to the image bundle. Upload an image. Make sure the set alt text (not a decorative image.)
  d. Save the node.
  e. Use this metatag checking tool to inspect the tag and see a social preview. Verify that the image, meta title, and description match the content entered in this test plan. [https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fcomponents%2Fhero-images-full-width-region-and-main-region%2Fshort-spotlights-classic](https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fcomponents%2Fhero-images-full-width-region-and-main-region%2Fshort-spotlights-classic)
  f. Alternatively, you can view the page source. The <head> of the page will include something similar to:

```html
<meta name="description" content="Here is my custom summary" />
...
<meta property="og:image" content="https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com/sites/hs_colorful/files/styles/metatags_image_1200x628/public/media/image/pexels-pixabay-36717_0.jpg.webp?itok=YJm8p1V-" />
<meta property="og:image:alt" content="my alt text" />
<meta name="twitter:image" content="https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com/sites/hs_colorful/files/styles/metatags_image_1200x628/public/media/image/pexels-pixabay-36717_0.jpg.webp?itok=YJm8p1V-" />
<meta name="twitter:image:alt" content="my alt text" />
...
<title>My custom title</title>
```

4. Test fallbacks
  a. Edit the node again. Remove all content from the meta fields. No title, description or image.
  b. Make sure that this node still has a banner image in the full width region.
  c. Save the node.
  d. Use this metatag checking tool to inspect the tag and see a social preview. Verify that the image, meta title, and description match the content entered in this test plan. [https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fcomponents%2Fhero-images-full-width-region-and-main-region%2Fshort-spotlights-classic](https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fcomponents%2Fhero-images-full-width-region-and-main-region%2Fshort-spotlights-classic)
  e. Verify that the title falls back to `Short Spotlight(s) - Classic | Center for Website Studies`, the description is empty, and that the meta image uses the first spotlight on the page as a fallback.
5. Test fallback summary for a non-page.
  a. Check an event node (since this does not have custom meta fields) and vierify that the node still has a description [https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fevents%2Fsociology-department-colloquium-robert-braun](https://metatags.io/?url=https%3A%2F%2Fpr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com%2Fevents%2Fsociology-department-colloquium-robert-braun)
6. Code review this contrib module at 1.0.0 release: https://git.drupalcode.org/project/metatag_simple_widget/-/tree/1.0.0?ref_type=tags

<img width="360" height="768" alt="meta-fields" src="https://github.com/user-attachments/assets/9c7b8e43-f5ca-46a0-93b4-0697cd868cc7" />

# Review Tasks

## Backend / Functional Validation
### Code
- [ ] Are the naming conventions following our standards?
- [ ] Are PHP functions and variables in `snake_case` and not `camelCase`?
- [ ] Does Drupal code follow [Drupal Coding Standards](https://www.drupal.org/docs/develop/standards/php/php-coding-standards)?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided?

### Code security
- [ ] Is all [user input properly sanitized when rendered](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

### General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210814338594933
  - https://app.asana.com/0/0/1211009229686701

nJim committed:
feat(SHS-6200): set metatag defaults and define new image stlye

nJim committed:
feat(SHS-6200): set metatag defaults and define new image stlye

nJim committed:
feat(SHS-6199): enable metatag simple widget on pages

nJim committed:
feat(SHS-6199): place metatag field with update hook

nJim committed:
feat(SHS-6199): merge lates from 11.12.1-release

nJim committed:
Merge branch '11.12.1-release' into SHS-6200-mtatag-defaults

nJim committed:
Merge branch 'SHS-6200-mtatag-defaults' into SHS-6199-mtags-widget

nJim committed:
feat(SHS-6199): lint fix

nJim committed:
feat(SHS-6201): add image field for metatags

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-colorful-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-traditional-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://english-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://history-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://west-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231

Visual Diffs:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231/visualdiffs

nJim committed:
fix(SHS-6201): allow page meta image to fallback on banner

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-colorful-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-traditional-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://english-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://history-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://west-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231

Visual Diffs:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231/visualdiffs

cienvaras committed:
Merge branch '11.12.2-release' of github.com:SU-HSDO/suhumsci into SHS-6201-meta-images

cienvaras committed:
fix(shs-6201): add missing composer.lock

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-colorful-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-traditional-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://english-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://history-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://west-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231

Visual Diffs:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231/visualdiffs

cienvaras reviewed (approved):
@nJim LGTM, thanks!

@ahughes3 Ready for you

ahughes3 commented:
@nJim The metatag settings are no longer available when editing or creating an Event node, is that the expected behavior for this new PR?

ahughes3 commented:
@cienvaras @nJim I attempted to rebuild the tugboat and it failed, I believe we need to first resolve the conflicts on the PR

nJim committed:
feat(SHS-6201): merge release

nJim committed:
feat(SHS-6201): merge release

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-colorful-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-traditional-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://english-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://history-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://west-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231

Visual Diffs:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231/visualdiffs

nJim commented:
@ahughes3 Composer lock file rebuild and merge conflict resolved. No functional changes. It is good to go.

Jnwibowo7321 reviewed (approved):
Tested on Tugboat. LGTM

joegl reviewed (commented):
I noticed that this PR adds new fields and their field storage via config, but also includes update hooks (su_humsci_profile_update_9734/9735) that add those fields to the entity form display. Since update hooks run before config import during deployment, these hooks may execute before the fields exist, which could cause the form display config to be updated for fields that aren’t present yet.

Is there a reason it's being done this way?

Also I noticed there a couple conflicts to clean up before this can be merged. Otherwise it's looking pretty good 👍 

Thanks!

ahughes3 commented:
@cienvaras @nJim This PR has merge conflicts that need to be resolved for Tugboat to regenerate a preview

nJim committed:
fix(SHS-6201): merge release branche

nJim committed:
fix(SHS-6201): merge release branch

nJim committed:
fix(SHS-6201): merge release branch

nJim committed:
fix(SHS-6201): enable metatag_simple_widget via update hook

nJim commented:
@joegl Thank you for pointing this out. I'm not accustom to working with sites where the form is config ignored. I fixed this with a more comprehensive update hook -- first enabling the new module, then manually creating the fields, and finally placing the fields on the form. This follows a previously used pattern in the profile update hooks.

@ahughes3 I updated the merge target and resolved the conflicts. 

Codeception is failing, but I believe this is unrelated to this change-set. I believe this work is again ready for review.

Jnwibowo7321 commented:
Thank you @nJim, @joegl ready for your review again for merging.

joegl committed:
Merge branch '11.13.1-release' into SHS-6201-meta-images

joegl commented on docroot/profiles/humsci/su_humsci_profile/su_humsci_profile.install:
```diff
@@ -1553,3 +1553,114 @@ function su_humsci_profile_update_9736() {
+ * Add field_metatags to hs_basic_page entity form display.
+ */
+function su_humsci_profile_update_9737() {
+  // Install required module.
+  \Drupal::service('module_installer')->install(['metatag_simple_widget']);
```
I'd put this in it's own update hook before the field config changes.

joegl reviewed (changes requested):
The `entity_form_display` configuration for `hs_basic_page` node is missing the two metatag fields added here. I know there is a `config_ignore` rule for `entity_form_display` configurations on nodes, but we still need the default configuration updated for new site installs, and for local site syncs. 

Otherwise this looks good and is working as expected.

cienvaras committed:
Merge branch '11.13.1-release' of github.com:SU-HSDO/suhumsci into SHS-6201-meta-images

cienvaras committed:
fix(shs-6201): move metatag_simple_widget module install to its own update hook

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1897-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-colorful-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://hs-traditional-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://english-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://history-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com
- https://west-0ysotll3mbc0qybuhqa3txhprbqkqikw.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231

Visual Diffs:
- https://dashboard.tugboatqa.com/6876ad23732855b6d5ddf231/visualdiffs

cienvaras committed:
fix(shs-6201): update basic page form and view display updated config

cienvaras commented:
@joegl 

> The entity_form_display configuration for hs_basic_page node is missing the two metatag fields added here. I know there is a config_ignore rule for entity_form_display configurations on nodes, but we still need the default configuration updated for new site installs, and for local site syncs.

Fixed, thanks. Also moved the module install to a separate update hook. It's ready for review again.

joegl commented:
> @joegl
> 
> > The entity_form_display configuration for hs_basic_page node is missing the two metatag fields added here. I know there is a config_ignore rule for entity_form_display configurations on nodes, but we still need the default configuration updated for new site installs, and for local site syncs.
> 
> Fixed, thanks. Also moved the module install to a separate update hook. It's ready for review again.

Thanks @cienvaras !
